### PR TITLE
Separate source container exit task

### PR DESCRIPTION
**Pull request by tkdchen**
* COULDBLD-8272

Split the exit plugin definition into a separate exit task, and always
execute this exit task as the last step in the pipeline.

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- n/a JSON/YAML configuration changes are updated in the relevant schema
- n/a Changes to metadata also update the documentation for the metadata
- n/a Pull request has a link to an osbs-docs PR for user documentation updates
- n/a New feature can be disabled from a configuration file

### CODE REVIEW

File: atomic_reactor/cli/parser.py [hunk 1]
```diff
@@ -50,12 +50,19 @@ def parse_args(args: Optional[Sequence[str]] = None) -> dict:
...
         help="build a source container",
         description="Build a source container.",
     )
-    source_build.set_defaults(func=task.source_build)
+    source_container_build.set_defaults(func=task.source_container_build)
+
+    source_container_exit = tasks.add_parser(
+        "source-container-exit",
+        help="Source container build exit task.",
+        description="Run plugins while exiting a source container build.",
```
Comment (chmeliik) [inline atomic_reactor/cli/parser.py:63]: Could this be consistent with the binary container exit task?
```suggestion
        help="exit a source container build",
        description="Execute source container exit steps.",
```

File: atomic_reactor/cli/task.py [hunk 2]
```diff
@@ -10,7 +10,9 @@
 from atomic_reactor.tasks.clone import CloneTask
 from atomic_reactor.tasks.common import TaskParams
 from atomic_reactor.tasks.orchestrator import OrchestratorTask, OrchestratorTaskParams
-from atomic_reactor.tasks.sources import SourceBuildTask, SourceBuildTaskParams
+from atomic_reactor.tasks.sources import (
+    SourceBuildExitTask, SourceBuildPhasesTask, SourceBuildTaskParams
```
Comment (chmeliik) [inline atomic_reactor/cli/task.py:14]: Again for consistency with binary tasks, maybe the classes should be `SourceExitTask`, `SourceBuildTask`, `SourceTaskParams`?

File: atomic_reactor/cli/task.py [hunk 3]
```diff
@@ -34,13 +36,23 @@ def worker(task_args: dict):
...
+    :param task_args: CLI arguments for a source-container-build task
     """
     params = SourceBuildTaskParams.from_cli_args(task_args)
-    task = SourceBuildTask(params)
+    task = SourceBuildPhasesTask(params)
+    return task.execute()
+
+
+def source_container_exit(task_args: dict):
+    """Run a source container build.
```
Comment (chmeliik) [inline atomic_reactor/cli/task.py:50]:
```suggestion
    """Run source container exit steps.
```

File: atomic_reactor/tasks/sources.py [hunk 4]
```diff
@@ -44,15 +53,17 @@ class SourceBuildTask(plugin_based.PluginBasedTask):
...
+    """Source container build task."""
+
+    plugins_def = plugin_based.PluginsDef(
         exit=[
             {"name": "verify_media", "required": False},
             {"name": "koji_import_source_container"},
             {"name": "koji_tag_build"},
             {"name": "store_metadata"},
         ],
     )
```
Comment (chmeliik) [inline atomic_reactor/tasks/sources.py:69]: This won't work, because the `workflow.build_docker_image()` method cannot handle when there are 0 buildstep plugins.

@mkosiarc, you encountered this while working on the build task, do you have a solution yet? If yes, could it be a standalone PR?

Comment (mkosiarc) [reply atomic_reactor/tasks/sources.py:69]: I don't have it yet, but will work on it soon. If I finish this and test that it works, I can create a standalone PR.

Comment (chmeliik) [reply atomic_reactor/tasks/sources.py:69]: Ack, just to make sure @tkdchen and you don't do the same work

Comment (chmeliik) [reply atomic_reactor/tasks/sources.py]: It won't work even if you move the plugins from `exit` to `prebuild` :disappointed: 

At least the plugins will get executed this way, but after that the finally task will fail with "No appropriate build step"

File: tekton/tasks/source-container-exit.yaml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+  workspaces:
+    - name: ws-build-dir
+    - name: ws-context-dir
+    - name: ws-registries-secret  # access with $(workspaces.ws-registries-secret.path)/token
+    - name: ws-koji-secret  # access with $(workspaces.ws-koji-secret.path)/token
+    - name: ws-reactor-config-map
+
+  steps:
+    - name: source-container-exit-image
```
Comment (chmeliik) [inline tekton/tasks/source-container-exit.yaml:24]:
```suggestion
    - name: source-container-exit
```

File: tekton/tasks/source-container-exit.yaml [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+      image: $(params.OSBS_IMAGE)
+      resources:
+        requests:
+          memory: 3000Mi
+          cpu: 500m
+        limits:
+          memory: 8600Mi
+          cpu: 1300m
```
Comment (chmeliik) [inline tekton/tasks/source-container-exit.yaml:32]: This could probably use the same resources as the binary container exit task. Only the source-container-build task should need the extra resources

**Comment by tkdchen** [comment]
This pr has to be merged after https://github.com/containerbuildsystem/atomic-reactor/pull/1821.

File: atomic_reactor/tasks/sources.py [hunk 7]
```diff
@@ -44,15 +53,17 @@ class SourceBuildTask(plugin_based.PluginBasedTask):
...
         ],
-        exit=[
+    )
+
+
+class SourceExitTask(SourceBuildBaseTask):
+    """Source container exit task."""
+
+    plugins_def = plugin_based.PluginsDef(
+        prebuild=[
```
Comment (chmeliik) [inline atomic_reactor/tasks/sources.py:63]: Could this be `exit` now?

I think there is some logic in `plugin.py` which makes it so that the `PrebuildPluginsRunner` will refuse to run exit plugins.

Comment (tkdchen) [reply atomic_reactor/tasks/sources.py]: Done. It's now `exit`.

**Review by chmeliik — APPROVED** [review]
LGTM :+1:

File: tekton/pipelines/source-container.yaml [hunk 8]
```diff
@@ -38,3 +38,23 @@ spec:
           value: "$(params.OSBS_IMAGE)"
         - name: USER_PARAMS
           value: "$(params.USER_PARAMS)"
+  finally:
+    - name: source-container-exit
```
Comment (hjmodi) [inline tekton/pipelines/source-container.yaml:42]: the name of the build task above(https://github.com/containerbuildsystem/atomic-reactor/pull/1813/files#diff-37fd879ef818700016f8f18b08a45989f6a7eaa79c3090b68180c1d541a1402cR22) should also be consistent with this naming

**Comment by hjmodi** [comment]
LGTM. Just one minor comment

File: tekton/pipelines/source-container.yaml (hunk 8 above)
Comment (tkdchen) [reply tekton/pipelines/source-container.yaml:42]: Update, PTAL.

**Review by hjmodi — APPROVED** [review]

**Comment by tkdchen** [comment]
Rebased.
